include/contraction/contractionsHierarchy.hpp

```diff
@@ -0,0 +1,246 @@
+                    Identifiers<int64_t> r;
+                    r.set_ids(reached_vertices_ids);
+                    log << "    >>> Reached vertices: " << r << std::endl;
+                }
+                catch ( boost::exception const &except ) {
+                    log << std::endl << "Boost exception during vertex contraction! " << dynamic_cast<std::exception const &>(except).what() << std::endl;
+                }
+                catch ( ... ) {
+                    log << "    Unknown exception during labelling!" << std::endl; 
+                }
+                /* abort in case of an interruption occurs (e.g. the query is being cancelled) */
+                CHECK_FOR_INTERRUPTS();                    
+                // Create a shortcut, for each w, when c(u, v) + c(u, w) > cost(shortest path from u to v)
+                for (const auto &w : adjacent_out_vertices) {
+                    E e, f, g;
+                    bool found_e, found_f, found_g;
+                    double c;                        
+                    if ( u != w ) {
+                        boost::tie(e, found_e) = boost::edge(u, v, graph.graph);
+                        boost::tie(f, found_f) = boost::edge(v, w, graph.graph);
+                        boost::tie(g, found_g) = boost::edge(u, w, graph.graph);
+                        if (found_e && found_f && (!found_g || (found_g && (distances[w] < graph[g].cost)))) {
+                            c = graph[e].cost + graph[f].cost;
+                            if ((distances[w] <= c) && (graph.is_shortcut_possible(u, v, w))) { 
+                                if (!simulation) {
+                                    pgrouting::CH_edge ch_e = graph.process_shortcut_from_shortest_path(u, v, w, distances[w], log);
+                                    shortcuts.push_back(ch_e);
+                                }
+                                shortcuts_number++;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        if (!simulation) {
+            for (auto &w : adjacent_out_vertices)
+                boost::remove_edge(v, w, graph.graph);
+            for (auto &u : graph.find_adjacent_in_vertices(v))
+                boost::remove_edge(u, v, graph.graph);
+            (graph[v]).clear_contracted_vertices();
+            log << "  Size of the graph after contraction: " << num_vertices(graph.graph) << " vertices and " << num_edges(graph.graph) << " edges" << std::endl;
+            log << "  " << shortcuts_number << " shortcuts created, " << old_edges << " old edges" << std::endl;
+        }
+        log << "  Metric: edge difference = " << shortcuts_number - old_edges << std::endl;
+        return static_cast<double>(shortcuts_number - old_edges);
+    }
+
+    void doContraction(
+        G &graph, 
+        std::ostringstream &log
+    )
+    {
+        // First iteration over vertices
+        G graph_copy=graph;
```

Comment:
```suggestion
        G graph_copy = graph;
```

You must use the pgrouting cpp formater if it exists